# <img style="float: left; padding-right: 20px; width: 200px" src="https://raw.githubusercontent.com/raxlab/imt2200-data/main/media/logo.jpg">  IMT 2200 - Introducción a Ciencia de Datos
**Pontificia Universidad Católica de Chile**<br>
**Instituto de Ingeniería Matemática y Computacional**<br>
**Semestre 2025-S2**<br>
**Profesor:** Rodrigo A. Carrasco <br>

# <h1><center>Actividad 05: Limpieza y Transformación de Datos</center></h1>

Esta actividad busca aplicar los conocimientos en transformación de datos para contestar algunas preguntas.

## Instrucciones

Esto Notebook contiene las instrucciones a realizar para la actividad. 

<b>Al finalizarla, deben subir el Notebook y los archivos generados en un único archivo .zip, al módulo de la Actividad 05 en Canvas.</b>

Para esta actividad trabajarán con el archivo `spotify_history.csv`, que contiene el historial de actividad de un usuario de Spotify durante varios años.

## Rúbrica

- Si han logrado todo (con errores menores): 7.0
- Si han logrado los puntos 1 y 2 (con errores menores): 5.0
- Si no han logrado hasta el punto 2.0: 1.0


## 1. Lectura y selección de columnas

a. Primero, lea el `spotify_history.csv` con Pandas como un único DataFrame y muestre todas las columnas del archivo original.

b. Para esta actividad, las columnas de nuestro interés son las siguientes:
```
['ts','ms_played',
'master_metadata_track_name', 'master_metadata_album_artist_name',
'master_metadata_album_album_name', 'episode_name',
'episode_show_name',
'reason_start', 'reason_end']
```
Seleccione solamente estas columnas para su DataFrame de trabajo.

c. Responda:
* ¿Cuántas filas tiene su dataset?

In [2]:
# a.
import pandas as pd


In [5]:
# b.
df = pd.read_csv('data2025/spotify_history.csv')
df.columns


Index(['ts', 'platform', 'ms_played', 'conn_country',
       'master_metadata_track_name', 'master_metadata_album_artist_name',
       'master_metadata_album_album_name', 'spotify_track_uri', 'episode_name',
       'episode_show_name', 'spotify_episode_uri', 'audiobook_title',
       'audiobook_uri', 'audiobook_chapter_uri', 'audiobook_chapter_title',
       'reason_start', 'reason_end', 'shuffle', 'offline', 'offline_timestamp',
       'incognito_mode'],
      dtype='object')

In [ ]:
# c.
columnas = ['ts','ms_played',
'master_metadata_track_name', 'master_metadata_album_artist_name',
'master_metadata_album_album_name', 'episode_name',
'episode_show_name',
'reason_start', 'reason_end']

df = df[columnas]
df_spotify = df.copy()
df_spotify = df_spotify[columnas]
df_spotify.head()

,ts,ms_played,master_metadata_track_name,master_metadata_album_artist_name,master_metadata_album_album_name,episode_name,episode_show_name,reason_start,reason_end
0,2022-12-29T16:39:18Z,753,Warning Sign,Coldplay,A Rush of Blood to the Head,NaN,NaN,fwdbtn,fwdbtn
1,2022-12-29T16:39:19Z,1661,Houdini,Foster The People,Torches,NaN,NaN,fwdbtn,fwdbtn
2,2022-12-29T16:39:21Z,1000,Fingerprints,Hiatus Kaiyote,Choose Your Weapon,NaN,NaN,fwdbtn,fwdbtn
3,2022-12-29T16:39:22Z,1097,Waves of Nature,WILLOW,ARDIPITHECUS,NaN,NaN,fwdbtn,backbtn
4,2022-12-29T16:43:38Z,256920,Fingerprints,Hiatus Kaiyote,Choose Your Weapon,NaN,NaN,backbtn,trackdone


## 2. Limpieza


a. Revise si existen filas duplicadas en su dataset. Si es así, deje solamente la última aparición de la fila duplicada en su DataFrame. ¿Cuántas filas duplicadas hay?

b. Ahora, revise si hay filas cuyo valor en `ms_played` (milisegundos reproducidos) es menor a 1. Descarte todas estas filas.

c. ¿Cuántas filas quedaron en su DataFrame después de la limpieza?

In [ ]:
# a.
duplicados = df.duplicated().sum()

print(f"Hay un total de {duplicados} filas duplicadas en el DataFrame.")
df_spotify = df_spotify.drop_duplicates(keep='last')

Hay un total de 71 filas duplicadas en el DataFrame.


In [17]:
# b.
menores_a_uno = (df['ms_played'] < 1).sum()
print("Número de filas con ms_played < 1:", menores_a_uno)

df_spotify = df_spotify[df_spotify['ms_played'] >= 1]

Número de filas con ms_played < 1: 271


In [18]:
# c.
print("Número de filas después de la limpieza:", len(df_spotify))

Número de filas después de la limpieza: 41487


## 3. Transformación

Aplique las siguientes transformaciones a su DataFrame:

a. La columna `ts` (timestamp) muestra la fecha y hora de reproducción de la canción. Extraiga únicamente el año y agregue una nueva columna con este dato como número entero. ¿En qué año hubo más reproducciones?

b. El dataset contiene canciones y episodios de podcasts. Genere una nueva columna (categórica o booleana) que identifique a cada fila como "canción" o "episodio" y explique cómo realizó esta clasificación. ¿Cuántas filas hay de cada una de estas categorías? 

c. Responda:
¿Cuál es el podcast con mayor número de reproducciones en total?

In [21]:
# a.
df_spotify['ts'] = pd.to_datetime(df_spotify['ts'], errors='coerce')
df_spotify['year'] = df_spotify['ts'].dt.year
reproduccion_año = df_spotify['year'].value_counts()
print(reproduccion_año)

year
2024    18860
2023    15426
2025     7079
2022      122
Name: count, dtype: int64


El año con más reproducciones fue XXX.

In [ ]:
# b.

In [22]:
# c.
podcast = df_spotify['episode_show_name'].value_counts()
print(podcast.head(1))

episode_show_name
Lateral with Tom Scott    211
Name: count, dtype: int64


El podcast con más reproducciones es Lateral with Tom Scott, 211 YYY reproducciones.

## 4. Variables categóricas

a. Las columnas `reason_start`y `reason_end` muestran, respectivamente, por qué se comenzó y se terminó la reproducción de una pista. Estas son columnas categóricas. Para ambas, muestre todas las categorías que aparecen en su dataset.
¿Son las mismas? Comente.

b. Vamos a crear una nueva columna llamada `skipped` (categórica o booleana), que identificará si la reproducción de la pista fue interrumpida intencionalmente por el usuario (saltada). Para esto, nos fijaremos en el valor de `reason_end` de cada fila. Considere una pista como "skipped" si su término está entre las siguientes categorías: `endplay`, `fwdbtn`, `backbtn`, y como "not skipped" en caso contrario.

c. Finalmente, responda:
¿Cuál es la canción más interrumpida (o saltada)? Responda con el nombre de la canción y el artista.

In [ ]:
# a.

In [ ]:
# b.

In [ ]:
# c.

La canción más interrumpida por el usuario es XXX.